Sources/NIOExtras/LineBasedFrameDecoder.swift

```diff
@@ -0,0 +1,97 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// A decoder that splits incoming `ByteBuffer`s around line end
+/// character(s) (`'\n'` or `'\r\n'`).
+///
+/// Let's, for example, consider the following received buffer:
+///
+///     +----+-------+------------+
+///     | AB | C\nDE | F\r\nGHI\n |
+///     +----+-------+------------+
+///
+/// A instance of `LineBasedFrameDecoder` will split this buffer
+/// as follows:
+///
+///     +-----+-----+-----+
+///     | ABC | DEF | GHI |
+///     +-----+-----+-----+
+///
+public class LineBasedFrameDecoder: ByteToMessageDecoder {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public var cumulationBuffer: ByteBuffer?
+    // keep track of the last scan offset from the buffer's reader index (if we didn't find the delimiter)
+    private var lastScanOffset: Int? = nil
+    
+    public init() { }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        if let frame = try self.findNextFrame(buffer: &buffer) {
+            ctx.fireChannelRead(wrapInboundOut(frame))
+            return .continue
+        } else {
+            return .needMoreData
+        }
+    }
+    
+    private func findNextFrame(buffer: inout ByteBuffer) throws -> ByteBuffer? {
+        var view = buffer.readableBytesView
+        // get the view's true start, end indexes
+        let _startIndex = view.startIndex
+        // start where we left off or from the beginning
+        let lastIndex = self.lastScanOffset.flatMap { buffer.readerIndex + $0 }
+        let firstIndex = min(lastIndex ?? _startIndex, view.endIndex)
+        view = view.dropFirst(firstIndex - buffer.readerIndex)
```

Comment:
Looks like we're doing some unnecessary math above. We're persisting the offset from the reader index, extracting the new reader index, doing two optionality checks, and a `min` function, and then subtracting the reader index again, all to answer the question "how far past the reader index should we go?"

However, we already knew the answer: it's `self.lastScanOffset ?? 0`. So lines 53 through 59 can be replaced by `var view = buffer.readableBytesView.dropFirst(self.lastScanOffset ?? 0)`.

Reply:
Oh in fact, we don't even need the optional folding, because we can now make `lastScanOffset` non-optional and just set it to `0` whenever we don't have one.